### ISSUE TITLE
Replace Windows stdin-wrapper executable with something better

### ISSUE THREAD

**Issue by hauleth** [w0rp/ale#117]

**Comment by w0rp** [comment]
I actually have another plan for this. I plan to eventually add support for chaining commands, so you can use the output of one command run asynchronously to run another job asynchronously. Then the need for special D script would go away. As far as the application of the idea goes, it would look like this:

```vim
function! GetDUBCommand(buffer)
    return 'dub describe'
endfunction

function! GetDMDCommand(buffer, dub_output_lines)
    " Build import paths arguments here
    return 'dmd ...'
endfunction

ale#linters#Define('d', {
\ ...
\   'command_chain' ['GetDUBCommand', 'GetDMDCommand'],
\})
``` 

The implementation is another story. That can be quite complicated.

**Comment by bounceme** [comment]
What do you think about https://github.com/w0rp/ale/issues/16#issuecomment-254011781 ?
I can see about making a pr to replace the stdin wrappers ( that is what this issue is about right? ) with a vimscript one

**Comment by w0rp** [comment]
This issue is about the DMD wrapper.  I don't think creating temporary files in the same directory as the source files is a good solution. Vim won't always manage to clean up the files after it is done, and you might not have permission to write the new files to the same directories.

**Comment by bounceme** [comment]
They are made in $TMPDIR . The calls of `delete()` clean the files out

**Comment by w0rp** [comment]
I see, you're copying the files to the temporary directory, and running the commands on that. That won't work when you need to check for files in a particular directory, say for when you import particular source files, etc.

**Comment by bounceme** [comment]
True, the linters produce output based on the temps, and the output has the file names switched for the loclists. Your wrappers are a cleaner solution

**Comment by hauleth** [comment]
I meant not `dmd-wrapper` but `stdin-wrapper.d`. I haven't used D in ages so `dmd-wrapper` isn't my area of interest.

**Comment by w0rp** [comment]
Ah right, the D program for the Windows stdin-wrapper? It could potentially be replaced with a PowerShell script if it will work on most versions of Windows. I'll go with whatever actually works. The 32-bit executable should be working.

**Comment by hauleth** [comment]
Depends on which Windows do you want to support. http://serverfault.com/questions/615275/in-which-version-of-windows-is-powershell-installed-by-default <http://serverfault.com/questions/615275/in-which-version-of-windows-is-powershell-installed-by-default>

TL;DR If we support only W7 and up, then we are ready to go.
--

Łukasz Niemier
lukasz@niemier.pl

**Comment by w0rp** [comment]
I think it would be reasonable to only support Windows 7 and up. If you're still running Windows XP or Vista now, there's something wrong with you.

**Comment by hauleth** [comment]
Still there would be need for Windows testing machine :P

**Comment by w0rp** [comment]
I do have a virtual machine set up with the plugin I could test it with.

**Comment by w0rp** [comment]
I have replaced the executable file with a Batch script now. I tried writing a PowerShell script, but there are some Windows environments where PowerShell scripts will not execute, so it's not portable enough for my liking. The Batch script should work well enough, and can probably be improved upon.

**Comment by w0rp** [comment]
I have re-opened this issue. I just reverted switching to the Batch script, as it seems to have caused problems with linting particular files, and I'd rather have something which works for now.

I am going to consider a new plan for all operating systems. I shall see if it's possible to first write the buffer data to a file in a temporary directory _from vim_ and then run the commands for returning errors directly, using the temporary file name as an argument. This will require some tweaking and testing.

**Comment by bounceme** [comment]
https://gist.github.com/bounceme/a2cb5c341af1b33c48689f64c678235f#file-neomakefly-vim-L12

would work, making changes to the specific filename of course. Is this going to only run for linters that don't take stdin? this loses the ability to do project-wide lintings

**Comment by w0rp** [comment]
Currently, linters which don't read stdin send the Vim buffer to either a Bash script or an Windows program which create a temporary file from the input from stdin and then run the command given to either the script or Windows program as arguments.

I'm thinking of replacing that with first writing to a temporary file in a temporary directory from Vim, and then calling the original command directly with the temporary filename formatted in. That should hopefully work on every OS, so long as the step for writing the file isn't too costly.

**Comment by bounceme** [comment]
`writefile()` is built in. I would assume that vim is efficient at saving a file, after all...

**Comment by w0rp** [comment]
I have now replaced all of the wrapper scripts with a Vim implementation, which should work basically the same on all operating systems. In addition, the temporary filename can be passed to a command at any location in a string, or even multiple times.

Before: 

```vim
'command': ale#util#stdin_wrapper . '.py some-python-tool',
```

After: 

```vim
'command': 'some-python-tool %t',
```

There's a function for escaping strings so linters can avoid having strings interpreted in this manner. See `:help #ale#linter#Define()` for more information.
